internal/queue/queue.go

```diff
@@ -0,0 +1,101 @@
+package queue
+
+import (
+	"container/ring"
+)
+
+// Queue[T] is a segmented FIFO queue of Ts.
+type Queue[T any] struct {
+	head *ring.Ring
+	tail *ring.Ring
+	size int
+}
+
+// New creates a new instance of Queue[T].
+//   - size is the size of each Queue segment
+func New[T any](size int) *Queue[T] {
+	r := &ring.Ring{
+		Value: &segment[T]{
+			items: make([]*T, size),
+		},
+	}
+	return &Queue[T]{
+		head: r,
+		tail: r,
+	}
+}
+
+// Enqueue adds the specified item to the end of the queue.
+// If the current segment is full, a new segment is created.
+func (q *Queue[T]) Enqueue(item T) {
+	for {
+		r := q.tail
+		seg := r.Value.(*segment[T])
+
+		if seg.tail < len(seg.items) {
+			seg.items[seg.tail] = &item
+			seg.tail++
+			q.size++
+			return
+		}
+
+		// segment is full, can we advance?
+		if next := r.Next(); next != q.head {
+			q.tail = next
+			continue
+		}
+
+		// no, add a new ring
+		r.Link(&ring.Ring{
+			Value: &segment[T]{
+				items: make([]*T, len(seg.items)),
+			},
+		})
+
+		q.tail = r.Next()
+	}
+}
+
+// Dequeue removes and returns the item from the front of the queue.
+func (q *Queue[T]) Dequeue() *T {
+	r := q.head
+	seg := r.Value.(*segment[T])
+
+	if seg.tail == 0 {
+		// queue is empty
+		return nil
+	}
+
+	// remove first item
+	item := seg.items[seg.head]
+	seg.items[seg.head] = nil
+	seg.head++
+	q.size--
+
+	if seg.head == seg.tail {
+		// segment is now empty, reset indices
+		seg.head, seg.tail = 0, 0
+
+		// if we're not at the last ring, advance head to the next one
+		if next := r.Next(); next != q.head {
+			q.head = next
+		} else {
+			// this is the last segment in the ring and it's empty, so the
+			// entire queue is empty. reset head and tail to this segment.
+			q.head, q.tail = r, r
+		}
+	}
+
+	return item
+}
+
+// Len returns the total count of enqueued items.
+func (q *Queue[T]) Len() int {
+	return q.size
+}
+
+type segment[T any] struct {
+	items []*T
```

Comment:
I need to measure if this introduces unwanted GC pressure.  Originally, the channel held `Message` by value.  I switched to byref to reduce the size of the array (`Message` is 168 bytes).  Need to confirm this doesn't make things worse.

Reply:
```
with *T
PS C:\git\Azure\go-amqp> go test -bench=BenchmarkMessageQueue -benchmem
goos: windows
goarch: amd64
pkg: github.com/Azure/go-amqp
cpu: AMD Ryzen 7 2700X Eight-Core Processor
BenchmarkMessageQueue-4              397           2947499 ns/op          959727 B/op       9775 allocs/op

with T
PS C:\git\Azure\go-amqp> go test -bench=BenchmarkMessageQueue -benchmem
goos: windows
goarch: amd64
pkg: github.com/Azure/go-amqp
cpu: AMD Ryzen 7 2700X Eight-Core Processor
BenchmarkMessageQueue-4              123           9839157 ns/op         1819882 B/op       9775 allocs/op
PASS
ok      github.com/Azure/go-amqp        26.619s
```